File: Source/Magick.NET.Native/Settings/MontageSettings.c
// Copyright 2013-2017 Dirk Lemstra <https://github.com/dlemstra/Magick.NET/>
//
// Licensed under the ImageMagick License (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
//
//   https://www.imagemagick.org/script/license.php
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific language governing permissions
// and limitations under the License.

#include "Stdafx.h"
#include "MontageSettings.h"

MAGICK_NET_EXPORT MontageInfo *MontageSettings_Create(void)
{
  MontageInfo
    *montage_info;

  ImageInfo
    *image_info;

  image_info=AcquireImageInfo();
  montage_info = AcquireMagickMemory(sizeof(*montage_info));
  GetMontageInfo(image_info, montage_info);
  DestroyImageInfo(image_info);
  return montage_info;
}

MAGICK_NET_EXPORT void MontageSettings_Dispose(MontageInfo *instance)
{
  DestroyMontageInfo(instance);
}

MAGICK_NET_EXPORT void MontageSettings_SetBackgroundColor(MontageInfo *instance, const PixelInfo *value)
{
  if (value != (PixelInfo*)NULL)
    instance->background_color = *value;
}

MAGICK_NET_EXPORT void MontageSettings_SetBorderColor(MontageInfo *instance, const PixelInfo *value)
{
  if (value != (PixelInfo*)NULL)
    instance->border_color = *value;
}

MAGICK_NET_EXPORT void MontageSettings_SetBorderWidth(MontageInfo *instance, const size_t value)
{
  instance->border_width = value;
}

MAGICK_NET_EXPORT void MontageSettings_SetFillColor(MontageInfo *instance, const PixelInfo *value)
{
  if (value != (PixelInfo*)NULL)
    instance->fill = *value;
}

MAGICK_NET_EXPORT void MontageSettings_SetFont(MontageInfo *instance, const char *value)
{
  CloneString(&instance->font, value);
}

MAGICK_NET_EXPORT void MontageSettings_SetFontPointsize(MontageInfo *instance, double value)
{
  instance->pointsize = value;
}

MAGICK_NET_EXPORT void MontageSettings_SetFrameGeometry(MontageInfo *instance, const char *value)
{
  CloneString(&instance->frame, value);
}

MAGICK_NET_EXPORT void MontageSettings_SetGeometry(MontageInfo *instance, const char *value)
{
  CloneString(&instance->geometry, value);
}

MAGICK_NET_EXPORT void MontageSettings_SetGravity(MontageInfo *instance, const size_t value)
{
  instance->gravity = (GravityType)value;
}

MAGICK_NET_EXPORT void MontageSettings_SetShadow(MontageInfo *instance, const MagickBooleanType value)
{
  instance->shadow = value;
}

MAGICK_NET_EXPORT void MontageSettings_SetStrokeColor(MontageInfo *instance, const PixelInfo *value)
{
  if (value != (PixelInfo*)NULL)
    instance->stroke = *value;
}

MAGICK_NET_EXPORT void MontageSettings_SetTextureFileName(MontageInfo *instance, const char *value)
{
  CloneString(&instance->texture, value);
}

MAGICK_NET_EXPORT void MontageSettings_SetTileGeometry(MontageInfo *instance, const char *value)
{
  CloneString(&instance->tile, value);
}

MAGICK_NET_EXPORT void MontageSettings_SetTitle(MontageInfo *instance, const char *value)
{
  CloneString(&instance->title, value);
}

File: src/Magick.NET/Shared/Settings/MontageSettings.cs
// Copyright 2013-2020 Dirk Lemstra <https://github.com/dlemstra/Magick.NET/>
//
// Licensed under the ImageMagick License (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
//
//   https://www.imagemagick.org/script/license.php
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific language governing permissions
// and limitations under the License.

#if Q8
using QuantumType = System.Byte;
#elif Q16
using QuantumType = System.UInt16;
#elif Q16HDRI
using QuantumType = System.Single;
#else
#error Not implemented!
#endif

namespace ImageMagick
{
    /// <summary>
    /// Class that contains setting for the montage operation.
    /// </summary>
    public sealed partial class MontageSettings
    {
        /// <summary>
        /// Gets or sets the color of the background that thumbnails are composed on.
        /// </summary>
        public IMagickColor<QuantumType> BackgroundColor { get; set; }

        /// <summary>
        /// Gets or sets the frame border color.
        /// </summary>
        public IMagickColor<QuantumType> BorderColor { get; set; }

        /// <summary>
        /// Gets or sets the pixels between thumbnail and surrounding frame.
        /// </summary>
        public int BorderWidth { get; set; }

        /// <summary>
        /// Gets or sets the fill color.
        /// </summary>
        public IMagickColor<QuantumType> FillColor { get; set; }

        /// <summary>
        /// Gets or sets the label font.
        /// </summary>
        public string Font { get; set; }

        /// <summary>
        /// Gets or sets the font point size.
        /// </summary>
        public int FontPointsize { get; set; }

        /// <summary>
        /// Gets or sets the frame geometry (width &amp; height frame thickness).
        /// </summary>
        public IMagickGeometry FrameGeometry { get; set; }

        /// <summary>
        /// Gets or sets the thumbnail width &amp; height plus border width &amp; height.
        /// </summary>
        public IMagickGeometry Geometry { get; set; }

        /// <summary>
        /// Gets or sets the thumbnail position (e.g. SouthWestGravity).
        /// </summary>
        public Gravity Gravity { get; set; }

        /// <summary>
        /// Gets or sets the thumbnail label (applied to image prior to montage).
        /// </summary>
        public string Label { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether drop-shadows on thumbnails are enabled or disabled.
        /// </summary>
        public bool Shadow { get; set; }

        /// <summary>
        /// Gets or sets the outline color.
        /// </summary>
        public IMagickColor<QuantumType> StrokeColor { get; set; }

        /// <summary>
        /// Gets or sets the background texture image.
        /// </summary>
        public string TextureFileName { get; set; }

        /// <summary>
        /// Gets or sets the frame geometry (width &amp; height frame thickness).
        /// </summary>
        public IMagickGeometry TileGeometry { get; set; }

        /// <summary>
        /// Gets or sets the montage title.
        /// </summary>
        public string Title { get; set; }

        /// <summary>
        /// Gets or sets the transparent color.
        /// </summary>
        public IMagickColor<QuantumType> TransparentColor { get; set; }

        private static string Convert(IMagickGeometry geometry)
        {
            if (geometry == null)
                return null;

            return geometry.ToString();
        }

        private INativeInstance CreateNativeInstance()
        {
            var instance = new NativeMontageSettings();
            instance.SetBackgroundColor(BackgroundColor);
            instance.SetBorderColor(BorderColor);
            instance.SetBorderWidth(BorderWidth);
            instance.SetFillColor(FillColor);
            instance.SetFont(Font);
            instance.SetFontPointsize(FontPointsize);
            instance.SetFrameGeometry(Convert(FrameGeometry));
            instance.SetGeometry(Convert(Geometry));
            instance.SetGravity(Gravity);
            instance.SetShadow(Shadow);
            instance.SetStrokeColor(StrokeColor);
            instance.SetTextureFileName(TextureFileName);
            instance.SetTileGeometry(Convert(TileGeometry));
            instance.SetTitle(Title);

            return instance;
        }
    }
}